File: leetcode/116-populating-next-right-pointers-in-each-node.js
// https://leetcode.com/problems/populating-next-right-pointers-in-each-node/ (top-interview-questions)

// Definition for a Node.
function Node(val, left, right, next) {
  this.val = val === undefined ? null : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
  this.next = next === undefined ? null : next;
};


/** constant extra space solution
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {

  let node = root;
  let nextLevelStart = root !== null ? root.left : null;

  while (nextLevelStart !== null) {
    
    // left -> right
    node.left.next = node.right;

    if (node.next !== null) {
      // right -> left ( of the parent neighboor node )
      node.right.next = node.next.left;
      node = node.next;
    } else {
      // we reach the end of the current level we must go to the start node of the next level
      node = nextLevelStart;
      nextLevelStart = nextLevelStart.left;
    }

  }

  return root;

};



/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
  const levels = [];
  const queue = root !== null ? [root] : [];

  while (queue.length > 0) {
    levels.push([]);
    let levelLength = queue.length;
    while (levelLength > 0) {
      const node = queue.pop();
      levels[levels.length - 1].push(node);
      if(node.left) queue.unshift(node.left);
      if(node.right) queue.unshift(node.right);
      levelLength -= 1;
    }
  }

  for (const level of levels) {
    for (let i = 0; i < level.length - 1; i++) {
      level[i].next = level[i+1];
    }
  }

  return root;
};

File: leetcode/116-populating-next-right-pointers-in-each-node.py
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node/ (top-interview-questions)


from typing import Optional


# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
    

class Solution:
    def connect(self, root: Optional[Node]) -> Optional[Node]:
        curr = root
        # pointer for the first node of the next level
        nxt_level = root.left if root else None

        while curr and nxt_level:
            # connect the left child with the right child
            curr.left.next = curr.right
            
            # connect the right child of the current node with the left child of the next neighbor
            if curr.next:
                curr.right.next = curr.next.left
            
            # visit the next neighboor
            curr = curr.next
            
            # if we reach the end of this level of nodes 
            # we should go to the starting node of the next level
            if not curr:
                curr = nxt_level
                nxt_level = curr.left
        
        
        return root
